core/src/main/java/com/linecorp/armeria/client/AutoUpdatingAddressResolver.java

```diff
@@ -0,0 +1,380 @@
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public long getDelay(TimeUnit unit) {
+            return Long.MIN_VALUE;
+        }
+
+        @Override
+        public int compareTo(Delayed o) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return true;
+        }
+
+        @Override
+        public boolean isDone() {
+            return true;
+        }
+
+        @Override
+        public Object get() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Object get(long timeout, TimeUnit unit) {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    private final ConcurrentMap<String, CompletableFuture<CacheEntry>> cache;
+    private final DefaultDnsNameResolver resolver;
+    private final List<DnsRecordType> dnsRecordTypes;
+    private final int minTtl;
+    private final int maxTtl;
+    private final Backoff autoUpdateBackoff;
+    private final long autoUpdateTimeoutMillis;
+
+    private volatile boolean resolverClosed;
+
+    AutoUpdatingAddressResolver(EventLoop eventLoop,
+                                ConcurrentMap<String, CompletableFuture<CacheEntry>> cache,
+                                DefaultDnsNameResolver resolver, List<DnsRecordType> dnsRecordTypes,
+                                int minTtl, int maxTtl, Backoff autoUpdateBackoff,
+                                long autoUpdateTimeoutMillis) {
+        super(eventLoop);
+        this.cache = cache;
+        this.resolver = resolver;
+        this.dnsRecordTypes = dnsRecordTypes;
+        this.minTtl = minTtl;
+        this.maxTtl = maxTtl;
+        this.autoUpdateBackoff = autoUpdateBackoff;
+        this.autoUpdateTimeoutMillis = autoUpdateTimeoutMillis;
+    }
+
+    @Override
+    protected boolean doIsResolved(InetSocketAddress address) {
+        return !requireNonNull(address, "address").isUnresolved();
+    }
+
+    @Override
+    protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocketAddress> promise)
+            throws Exception {
+        requireNonNull(unresolvedAddress, "unresolvedAddress");
+        requireNonNull(promise, "promise");
+        if (resolverClosed) {
+            promise.tryFailure(new IllegalStateException("resolver is closed already."));
+            return;
+        }
+        final String hostname = unresolvedAddress.getHostString();
+        final int port = unresolvedAddress.getPort();
+        final CompletableFuture<CacheEntry> entryFuture = cache.get(hostname);
+        if (entryFuture != null) {
+            handle(entryFuture, promise, port);
+            return;
+        }
+
+        final CompletableFuture<CacheEntry> result = new CompletableFuture<>();
+        final CompletableFuture<CacheEntry> previous = cache.putIfAbsent(hostname, result);
```

Comment:
I think DNS cache hit rate is very high. How about checking `previous` with `get` first to avoid additional allocation of `result`. 
Or we can split cache into result cache and pending work to concurrency. Does it seem overkill? WDYT? 😀

Reply:
It's up there. 😄

Reply:
I think caffeine provides a high performance, asynchronous cache that takes care of a lot of this logic (configuring it to refresh on interval is exactly the behavior we want of returning stale value while resolving new one). Have you thought of using it?

Reply:
Every cache has to have a different expiration timeout and it's very inconvenient to invalidate the cache. (Have to call `synchronous()`). And There were some other situations that make me hard to use it. So I decided not to use that. 😄

Reply:
I see - yeah it seems like it's missing value based refresh even though it has value based expiry. Filed issue - https://github.com/ben-manes/caffeine/issues/360